### PR TITLE
[Merged by Bors] - feat: add some constructions about `RelSeries` (`append`, `insert_nth`, `map`, `comap`)

### PR DESCRIPTION
**Pull request by jjaassoonn**
- appending a relation series to another one
- insert an element into a relation series to make a longer relation series
- map a relation series across a relation preserving function
- pullback a lt series across a surjective strictly co-monotonic function. (`f` is strictly co-monotonic if `f x < f y` implies `x < y`)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
See [zulip thread](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/StrictSeries.20!4.233858.2C.20!4.233852) for discussion about the design here.

File: Mathlib/Order/RelSeries.lean [hunk 1]
```diff
@@ -154,6 +157,148 @@ protected noncomputable def withLength [r.InfiniteDimensional] (n : ℕ) : RelSe
 lemma nonempty_of_infiniteDimensional [r.InfiniteDimensional] : Nonempty α :=
   ⟨RelSeries.withLength r 0 0⟩
 
+instance membership : Membership α (RelSeries r) :=
+  ⟨fun x s => x ∈ Set.range s⟩
```
Comment (YaelDillies) [inline Mathlib/Order/RelSeries.lean:161]:
```suggestion
  ⟨(· ∈ Set.range ·)⟩
```
if you're into this

File: Mathlib/Order/RelSeries.lean [hunk 2]
```diff
@@ -154,6 +157,148 @@ protected noncomputable def withLength [r.InfiniteDimensional] (n : ℕ) : RelSe
...
+
+theorem mem_def {x : α} {s : RelSeries r} : x ∈ s ↔ x ∈ Set.range s :=
+  Iff.rfl
+
+/-- start of a series -/
+def bot (x : RelSeries r) : α := x 0
+
+/-- end of a series -/
+def top (x : RelSeries r) : α := x <| Fin.last _
```
Comment (YaelDillies) [inline Mathlib/Order/RelSeries.lean:170]: I think this deserves slightly longer docstrings. Also those names are inconsistent with `List.head`, `Fin.head`, etc...

File: Mathlib/Order/RelSeries.lean [hunk 3]
```diff
@@ -154,6 +157,151 @@ protected noncomputable def withLength [r.InfiniteDimensional] (n : ℕ) : RelSe
...
+  ⟨(· ∈ Set.range ·)⟩
+
+theorem mem_def {x : α} {s : RelSeries r} : x ∈ s ↔ x ∈ Set.range s :=
+  Iff.rfl
+
+/-- start of a series .i.e. for `a₀ -r-> a₁ -r-> ... -r-> aₙ`, its head is `a₀`. Since a relation
+series is assumed to be non-empty, this is well defined. -/
+def head (x : RelSeries r) : α := x 0
+
+/-- end of a series .i.e. for `a₀ -r-> a₁ -r-> ... -r-> aₙ`, its head is `a¬`.  Since a relation
```
Comment (YaelDillies) [inline Mathlib/Order/RelSeries.lean:170]:
```suggestion
/-- End of a series, i.e. for `a₀ -r-> a₁ -r-> ... -r-> aₙ`, its last element is `a¬`.  Since a relation
```

File: Mathlib/Order/RelSeries.lean [hunk 4]
```diff
@@ -15,7 +17,8 @@ If `r` is a relation on `α` then a relation series of length `n` is a series
 
 -/
 
-variable {α : Type*} (r : Rel α α)
+variable {α : Type _} (r : Rel α α)
+variable {β : Type _} (s : Rel β β)
```
Comment (YaelDillies) [inline Mathlib/Order/RelSeries.lean:21]: Surely you mean `Type*`?
```suggestion
variable {α : Type*} (r : Rel α α)
variable {β : Type*} (s : Rel β β)
```

File: Mathlib/Order/RelSeries.lean [hunk 5]
```diff
@@ -154,6 +157,124 @@ protected noncomputable def withLength [r.InfiniteDimensional] (n : ℕ) : RelSe
...
   ⟨RelSeries.withLength r 0 0⟩
 
+instance membership : Membership α (RelSeries r) :=
+  ⟨(· ∈ Set.range ·)⟩
+
+theorem mem_def {x : α} {s : RelSeries r} : x ∈ s ↔ x ∈ Set.range s :=
+  Iff.rfl
+
+/-- Start of a series, i.e. for `a₀ -r-> a₁ -r-> ... -r-> aₙ`, its head is `a₀`. Since a relation
+series is assumed to be non-empty, this is well defined. -/
```
Comment (YaelDillies) [inline Mathlib/Order/RelSeries.lean:167]: For better hovers
```suggestion
/-- Start of a series, i.e. for `a₀ -r-> a₁ -r-> ... -r-> aₙ`, its head is `a₀`.

Since a relation series is assumed to be non-empty, this is well defined. -/
```

File: Mathlib/Order/RelSeries.lean [hunk 6]
```diff
@@ -154,6 +157,124 @@ protected noncomputable def withLength [r.InfiniteDimensional] (n : ℕ) : RelSe
...
+/--
+If `a_0 --r-> a_1 --r-> ... --r-> a_n` and `b_0 --r-> b_1 --r-> ... --r-> b_m` are two strict series
+such that `r a_n b_0`, then there is a chain of length `n + m + 1` given by
+`a_0 --r-> a_1 --r-> ... --r-> a_n --r-> b_0 --r-> b_1 --r-> ... --r-> b_m`.
+-/
+@[simps]
+def append (p q : RelSeries r) (connect : r p.last q.head) : RelSeries r where
+  length := p.length + q.length + 1
+  toFun := Fin.append p q ∘ Fin.cast (by abel)
+  step := fun i => by
```
Comment (YaelDillies) [inline Mathlib/Order/RelSeries.lean:187]:
```suggestion
  step i := by
```

File: Mathlib/Order/RelSeries.lean [hunk 7]
```diff
@@ -154,6 +157,124 @@ protected noncomputable def withLength [r.InfiniteDimensional] (n : ℕ) : RelSe
...
+          Nat.add_sub_cancel' <| le_of_lt (show p.length < i.1 from hi), add_comm]
+      rw [hx, Fin.append_right, hy, Fin.append_right]
+      convert q.step ⟨i - (p.length + 1), Nat.sub_lt_left_of_lt_add hi <|
+        by convert i.2 using 1; abel⟩
+      rw [Fin.succ_mk, Nat.sub_eq_iff_eq_add (le_of_lt hi : p.length ≤ i),
+        Nat.add_assoc _ 1, add_comm 1, Nat.sub_add_cancel]
+      exact hi
+
+/--
+For two sets `α, β` and relation on them `r, s`, if `f : α → β` preserves relation `r`, then an
```
Comment (YaelDillies) [inline Mathlib/Order/RelSeries.lean:213]:
```suggestion
For two types `α, β` and relation on them `r, s`, if `f : α → β` preserves relation `r`, then an
```

File: Mathlib/Order/RelSeries.lean [hunk 8]
```diff
@@ -154,6 +157,124 @@ protected noncomputable def withLength [r.InfiniteDimensional] (n : ℕ) : RelSe
...
+def map (p : RelSeries r)
+    (f : α → β) (hf : ∀ ⦃x y : α⦄, r x y → s (f x) (f y)) : RelSeries s where
+  length := p.length
+  toFun := f.comp p
+  step := (hf <| p.step .)
+
+/--
+If `a_0 --r-> a_1 --r-> ... --r-> a_n` is an `r`-series and `a` is such that
+`a_i --r-> a --r-> a_{i + 1}`, then
+`a_0 --r-> a_1 --r-> ... --r-> a_i --r-> a --r-> a_{i + 1} --r-> ... --r-> a_n`
```
Comment (YaelDillies) [inline Mathlib/Order/RelSeries.lean:227]: This notation is hard to parse. Can you either
* Use Unicode subscripts:
```suggestion
If `a₀ -r→ a₁ -r→ ... -r→ aₙ` is an `r`-series and `a` is such that
`aᵢ -r→ a -r→ a_ᵢ₊₁`, then
`a₀ -r→ a₁ -r→ ... -r→ a_i -r→ a -r→ aᵢ₊₁ -r→ ... -r→ aₙ`
```
or use latex?

File: Mathlib/Order/RelSeries.lean [hunk 9]
```diff
@@ -208,6 +329,35 @@ lemma strictMono (x : LTSeries α) : StrictMono x :=
 lemma monotone (x : LTSeries α) : Monotone x :=
   x.strictMono.monotone
 
+
+/-- an alternative constructor of `LTSeries` using `StrictMono` functions. -/
```
Comment (YaelDillies) [inline Mathlib/Order/RelSeries.lean:333]:
```suggestion
/-- An alternative constructor of `LTSeries` from a strictly monotone function. -/
```

File: Mathlib/Order/RelSeries.lean [hunk 10]
```diff
@@ -208,6 +329,35 @@ lemma strictMono (x : LTSeries α) : StrictMono x :=
...
+@[simps]
+def mk (length : ℕ) (toFun : Fin (length + 1) → α) (strictMono : StrictMono toFun) : LTSeries α :=
+{ toFun := toFun
+  step := fun i => strictMono <| lt_add_one i.1 }
+
+/--
+For two pre-ordered sets `α, β`, if `f : α → β` is strictly monotonic, then a strict chain of `α`
```
Comment (YaelDillies) [inline Mathlib/Order/RelSeries.lean:340]:
```suggestion
For two preorders `α, β`, if `f : α → β` is strictly monotonic, then a strict chain of `α`
```

File: Mathlib/Order/RelSeries.lean [hunk 11]
```diff
@@ -154,6 +157,125 @@ protected noncomputable def withLength [r.InfiniteDimensional] (n : ℕ) : RelSe
...
+def last (x : RelSeries r) : α := x <| Fin.last _
+
+lemma head_mem (x : RelSeries r) : x.head ∈ x := ⟨_, rfl⟩
+
+lemma last_mem (x : RelSeries r) : x.last ∈ x := ⟨_, rfl⟩
+
+/--
+If `a_0 --r-> a_1 --r-> ... --r-> a_n` and `b_0 --r-> b_1 --r-> ... --r-> b_m` are two strict series
+such that `r a_n b_0`, then there is a chain of length `n + m + 1` given by
+`a_0 --r-> a_1 --r-> ... --r-> a_n --r-> b_0 --r-> b_1 --r-> ... --r-> b_m`.
```
Comment (YaelDillies) [inline Mathlib/Order/RelSeries.lean:182]: Same here

File: Mathlib/Order/RelSeries.lean [hunk 12]
```diff
@@ -154,6 +157,125 @@ protected noncomputable def withLength [r.InfiniteDimensional] (n : ℕ) : RelSe
...
+      convert q.step ⟨i - (p.length + 1), Nat.sub_lt_left_of_lt_add hi <|
+        by convert i.2 using 1; abel⟩
+      rw [Fin.succ_mk, Nat.sub_eq_iff_eq_add (le_of_lt hi : p.length ≤ i),
+        Nat.add_assoc _ 1, add_comm 1, Nat.sub_add_cancel]
+      exact hi
+
+/--
+For two types `α, β` and relation on them `r, s`, if `f : α → β` preserves relation `r`, then an
+`r`-series can be pushed out to an `s`-series by
+`a₀ --r-> a₁ --r-> ... --r-> aₙ ↦ f a₀ --s-> f a₁ --s-> ... --s-> f aₙ`
```
Comment (YaelDillies) [inline Mathlib/Order/RelSeries.lean:216]: and there

File: Mathlib/Order/RelSeries.lean [hunk 13]
```diff
@@ -154,6 +157,125 @@ protected noncomputable def withLength [r.InfiniteDimensional] (n : ℕ) : RelSe
...
+          rw [Fin.insertNth_apply_above (h := hm)]
+          simp only [eq_rec_constant]
+          congr
+        · change Fin.insertNth _ _ _ _ = _
+          rw [Fin.insertNth_apply_above]
+          swap; exact hm.trans (lt_add_one _)
+          simp only [Fin.val_succ, Nat.zero_eq, Fin.pred_succ, eq_rec_constant, ge_iff_le,
+            Fin.succ_mk]
+          congr
+          exact Fin.ext <| Eq.symm <| Nat.succ_pred_eq_of_pos (lt_trans (Nat.zero_lt_succ _) hm)
```
Comment (YaelDillies) [inline Mathlib/Order/RelSeries.lean:269]: Any chance `aesop` can do some of those goals?

File: Mathlib/Order/RelSeries.lean [hunk 14]
```diff
@@ -154,6 +157,125 @@ protected noncomputable def withLength [r.InfiniteDimensional] (n : ℕ) : RelSe
...
+theorem mem_def {x : α} {s : RelSeries r} : x ∈ s ↔ x ∈ Set.range s :=
+  Iff.rfl
+
+/-- Start of a series, i.e. for `a₀ -r-> a₁ -r-> ... -r-> aₙ`, its head is `a₀`.
+
+Since a relation series is assumed to be non-empty, this is well defined. -/
+def head (x : RelSeries r) : α := x 0
+
+/-- End of a series, i.e. for `a₀ -r-> a₁ -r-> ... -r-> aₙ`, its last element is `aₙ`.  Since a
+relation series is assumed to be non-empty, this is well defined. -/
```
Comment (YaelDillies) [inline Mathlib/Order/RelSeries.lean:172]:
```suggestion
/-- Start of a series, i.e. for `a₀ -r→ a₁ -r→ ... -r→ aₙ`, its head is `a₀`.

Since a relation series is assumed to be non-empty, this is well defined. -/
def head (x : RelSeries r) : α := x 0

/-- End of a series, i.e. for `a₀ -r→ a₁ -r→ ... -r→ aₙ`, its last element is `aₙ`.  Since a
relation series is assumed to be non-empty, this is well defined. -/
```

File: Mathlib/Order/RelSeries.lean [hunk 15]
```diff
@@ -208,6 +330,35 @@ lemma strictMono (x : LTSeries α) : StrictMono x :=
...
+For two preorders `α, β`, if `f : α → β` is strictly monotonic, then a strict chain of `α`
+can be pushed out to a strict chain of `β` by
+`a₀ < a₁ < ... < aₙ ↦ f a₀ < f a₁ < ... < f aₙ`
+-/
+@[simps!]
+def map (p : LTSeries α) (f : α → β) (hf : StrictMono f) : LTSeries β :=
+  LTSeries.mk p.length (f.comp p) (hf.comp p.strictMono)
+
+/--
+For two pre-ordered sets `α, β`, if `f : α → β` is surjective and strictly comonotonic, then a
```
Comment (YaelDillies) [inline Mathlib/Order/RelSeries.lean:350]:
```suggestion
For two preorders `α, β`, if `f : α → β` is surjective and strictly comonotonic, then a
```

File: Mathlib/Order/RelSeries.lean [hunk 16]
```diff
@@ -154,6 +157,125 @@ protected noncomputable def withLength [r.InfiniteDimensional] (n : ℕ) : RelSe
...
+theorem mem_def {x : α} {s : RelSeries r} : x ∈ s ↔ x ∈ Set.range s :=
+  Iff.rfl
+
+/-- Start of a series, i.e. for `a₀ -r→ a₁ -r→ ... -r→ aₙ`, its head is `a₀`.
+
+Since a relation series is assumed to be non-empty, this is well defined. -/
+def head (x : RelSeries r) : α := x 0
+
+/-- End of a series, i.e. for `a₀ -r→ a₁ -r→ ... -r→ aₙ`, its last element is `aₙ`.  Since a
+relation series is assumed to be non-empty, this is well defined. -/
```
Comment (YaelDillies) [inline Mathlib/Order/RelSeries.lean:172]:
```suggestion
/-- End of a series, i.e. for `a₀ -r→ a₁ -r→ ... -r→ aₙ`, its last element is `aₙ`.

Since a relation series is assumed to be non-empty, this is well defined. -/
```

File: Mathlib/Order/RelSeries.lean [hunk 17]
```diff
@@ -154,6 +157,125 @@ protected noncomputable def withLength [r.InfiniteDimensional] (n : ℕ) : RelSe
...
+If `a₀ -r→ a₁ -r→ ... -r→ aₙ` and `b₀ -r→ b₁ -r→ ... -r→ bₘ` are two strict series
+such that `r aₙ b₀`, then there is a chain of length `n + m + 1` given by
+`a₀ -r→ a₁ -r→ ... -r→ aₙ -r→ b₀ -r→ b₁ -r→ ... -r→ bₘ`.
+-/
+@[simps]
+def append (p q : RelSeries r) (connect : r p.last q.head) : RelSeries r where
+  length := p.length + q.length + 1
+  toFun := Fin.append p q ∘ Fin.cast (by abel)
+  step i := by
+    obtain (hi|rfl|hi) :=
```
Comment (YaelDillies) [inline Mathlib/Order/RelSeries.lean:189]: You don't need those brackets.
```suggestion
    obtain hi | rfl | hi :=
```

File: Mathlib/Order/RelSeries.lean [hunk 18]
```diff
@@ -208,6 +330,35 @@ lemma strictMono (x : LTSeries α) : StrictMono x :=
 lemma monotone (x : LTSeries α) : Monotone x :=
   x.strictMono.monotone
 
+
+/-- An alternative constructor of `LTSeries` from a strictly monotone function. -/
+@[simps]
+def mk (length : ℕ) (toFun : Fin (length + 1) → α) (strictMono : StrictMono toFun) : LTSeries α :=
+{ toFun := toFun
+  step := fun i => strictMono <| lt_add_one i.1 }
```
Comment (YaelDillies) [inline Mathlib/Order/RelSeries.lean:338]:
```suggestion
def mk (length : ℕ) (toFun : Fin (length + 1) → α) (strictMono : StrictMono toFun) : LTSeries α where
  toFun := toFun
  step i := strictMono <| lt_add_one i.1
```

File: Mathlib/Order/RelSeries.lean [hunk 19]
```diff
@@ -154,6 +157,126 @@ protected noncomputable def withLength [r.InfiniteDimensional] (n : ℕ) : RelSe
...
+  toFun := Fin.append p q ∘ Fin.cast (by abel)
+  step i := by
+    obtain (hi|rfl|hi) :=
+      lt_trichotomy i (Fin.castLE (by linarith) (Fin.last _ : Fin (p.length + 1)))
+    · convert p.step ⟨i.1, hi⟩ <;> convert Fin.append_left p q _ <;> rfl
+    · convert connect
+      · convert Fin.append_left p q _; rfl
+      · convert Fin.append_right p q _; rfl
+    · set x := _; set y := _
+      change r (Fin.append p q x) (Fin.append p q y)
```
Comment (YaelDillies) [inline Mathlib/Order/RelSeries.lean:197]: Noting for later, can't you do something like
```suggestion
    · change r (Fin.append p q ?x) (Fin.append p q ?y)
```
I know this was talked about on Zulip but I don't know whether it was implemented.

**Review by YaelDillies — APPROVED** [review]
This is not perfect, but since you take ages to do every single change I point out (and not the similar ones until I point them out too), I'll call this good enough.

maintainer merge

File: Mathlib/Order/RelSeries.lean (hunk 19 above)
Comment (jjaassoonn) [reply Mathlib/Order/RelSeries.lean:197]: I just tried `change ... ?x`, but I don't think this works yet

**Comment by jjaassoonn** [comment]
> This is not perfect, but since you take ages to do every single change I point out (and not the similar ones until I point them out too), I'll call this good enough.
> 
> maintainer merge

Thank you for investing time in reviewing this PR, I realised that I haven't responded to the review carefully enough and in a timely manner. There is a similar PR #3852, which I will take my time to carefully apply suggestion already made here. I will try to make more effective response. Sorry for the trouble.

**Comment by YaelDillies** [comment]
That's okay! But indeed I would rather not waste much more time on minute details when the big picture is acceptable already.

**Comment by jjaassoonn** [comment]
> That's okay! But indeed I would rather not waste much more time on minute details when the big picture is acceptable already.

Noted. Thanks again for your patient review!

**Comment by kmill** [comment]
I see there's been a little more discussion since `maintainer merge`. Just to be sure in case there's anything else to do, here Yaël:

bors d=YaelDillies

**Comment by YaelDillies** [comment]
I think it's okay!

maintainer merge

**Comment by YaelDillies** [comment]
Sorry, wrong command.

bors merge
